src/storage-client/src/statistics.rs

```diff
@@ -18,50 +18,537 @@
+        // A gauge of the number of _values_ (source defined unit) we have committed.
+        // Never resets. Not to be confused with any of the counters above.
+        .with_column("offset_committed", ScalarType::UInt64.nullable(false))
+});
+
+pub static MZ_SINK_STATISTICS_RAW_DESC: Lazy<RelationDesc> = Lazy::new(|| {
+    RelationDesc::empty()
+        // Id of the sink.
+        .with_column("id", ScalarType::String.nullable(false))
+        //
+        // Counters
+        //
+        // A counter of the messages we have staged to upstream.
+        // Never resets.
+        .with_column("messages_staged", ScalarType::UInt64.nullable(false))
+        // A counter of the messages we have committed.
+        // Never resets.
+        .with_column("messages_committed", ScalarType::UInt64.nullable(false))
+        // A counter of the bytes we have staged to upstream.
+        // Never resets.
+        .with_column("bytes_staged", ScalarType::UInt64.nullable(false))
+        // A counter of the bytes we have committed.
+        // Never resets.
+        .with_column("bytes_committed", ScalarType::UInt64.nullable(false))
+});
+
+// Types of statistics (counter and various types of gauges), that have different semantics
+// when sources/sinks are reset.
+
+/// A counter that never resets.
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
+pub struct Counter(u64);
+
+impl Counter {
+    /// Merge `self` with another value.
+    /// Used to merge values across workers.
+    pub fn merge(&mut self, other: Self) {
+        // Sum across workers.
+        self.0 += other.0
+    }
+
+    /// Incorporate this value with another.
+    /// Used to update the values written to the builtin tables.
+    pub fn incorporate(&mut self, other: Self, _field_name: &'static str) {
+        // Always add the new value to the existing one.
+        self.0 += other.0
+    }
+}
+
+impl From<u64> for Counter {
+    fn from(f: u64) -> Self {
+        Counter(f)
+    }
+}
+
+/// A trait defining the reset semantics of various types of gauges.
+pub trait GaugeSemantics {
+    /// Merge `self` with another value.
+    /// Used to merge values across workers.
+    fn merge(&mut self, other: Self);
+
+    /// Incorporate this value with another.
+    /// Used to update the values written to the builtin tables.
+    fn incorporate(&mut self, other: Self, field_name: &'static str);
+}
+
+/// A latency gauge that is reset on every restart.
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
+pub struct ResettingLatency(Option<i64>);
+
+impl From<Option<i64>> for ResettingLatency {
+    fn from(f: Option<i64>) -> Self {
+        ResettingLatency(f)
+    }
+}
+
+impl GaugeSemantics for ResettingLatency {
+    fn merge(&mut self, other: Self) {
+        match (&mut self.0, other.0) {
+            (None, _) | (_, None) => {
+                // If any of the values are `NULL`, then we don't yet know the max.
+                self.0 = None
```

Comment:
What happens if the max latency arrives while the current value is `None`? Do we store it somewhere?

Reply:
yes, thats the `incorpoate`